### PR TITLE
Fix giving wrong column names when union has only right result.

### LINKED ISSUE #774: UNION return wrong num of cols
Screen Shot 2019-08-15 at 13 02 45

Screen Shot 2019-08-15 at 12 52 05

Left has two cols, Right has three cols
when the left return null, do right only, so the  cols is wrong

### PR DESCRIPTION
**Pull request by CPWstatic**
Fix #774

### CODE REVIEW

**Comment by CPWstatic** [comment]
Jenkins go!

**Comment by CPWstatic** [comment]
Jenkins go!

**Comment by CPWstatic** [comment]
Jenkins go!

File: src/graph/FetchEdgesExecutor.cpp [hunk 1]
```diff
@@ -109,13 +109,15 @@ Status FetchEdgesExecutor::setupEdgeKeysFromRef() {
     const InterimResult *inputs;
     if (sentence_->ref()->isInputExpr()) {
         inputs = inputs_.get();
-        if (inputs == nullptr) {
+        if (inputs == nullptr
+                || (inputs != nullptr && !inputs->hasData())) {
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: after ||, no need inputs != nullptr

File: src/graph/FetchEdgesExecutor.cpp [hunk 2]
```diff
@@ -109,13 +109,15 @@ Status FetchEdgesExecutor::setupEdgeKeysFromRef() {
...
             // we have empty imputs from pipe.
             return Status::OK();
         }
     } else {
         inputs = ectx()->variableHolder()->get(varname_);
-        if (inputs == nullptr) {
+        if (inputs == nullptr
+                || (inputs != nullptr && !inputs->hasData())) {
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: ditto

File: src/graph/FetchVerticesExecutor.cpp [hunk 3]
```diff
@@ -234,12 +234,14 @@ Status FetchVerticesExecutor::setupVidsFromRef() {
     const InterimResult *inputs;
     if (varname_ == nullptr) {
         inputs = inputs_.get();
-        if (inputs == nullptr) {
+        if (inputs == nullptr
+                || (inputs != nullptr && !inputs->hasData())) {
```
Comment (wadeliuyi) [inline src/graph/FetchVerticesExecutor.cpp]: ditto

File: src/graph/FetchVerticesExecutor.cpp [hunk 4]
```diff
@@ -234,12 +234,14 @@ Status FetchVerticesExecutor::setupVidsFromRef() {
...
             return Status::OK();
         }
     } else {
         inputs = ectx()->variableHolder()->get(*varname_);
-        if (inputs == nullptr) {
+        if (inputs == nullptr
+                || (inputs != nullptr && !inputs->hasData())) {
```
Comment (wadeliuyi) [inline src/graph/FetchVerticesExecutor.cpp]: ditto

File: src/graph/GoExecutor.cpp [hunk 5]
```diff
@@ -309,7 +309,8 @@ Status GoExecutor::setupStarts() {
         inputs = varInputs;
     }
     // No error happened, but we are having empty inputs
-    if (inputs == nullptr) {
+    if (inputs == nullptr
+            || (inputs != nullptr && !inputs->hasData())) {
```
Comment (wadeliuyi) [inline src/graph/GoExecutor.cpp]: ditto

File: src/graph/InterimResult.cpp [hunk 6]
```diff
@@ -13,23 +13,27 @@ namespace graph {
...
-    rsWriter_ = std::move(rsWriter);
-    rsReader_ = std::make_unique<RowSetReader>(rsWriter_->schema(), rsWriter_->data());
-}
-
-
 InterimResult::InterimResult(std::vector<VertexID> vids) {
     vids_ = std::move(vids);
 }
 
+InterimResult::InterimResult(std::vector<std::string> colNames) {
```
Comment (wadeliuyi) [inline src/graph/InterimResult.cpp]: maybe std::vector<std::string> &&colNames is better

File: src/graph/InterimResult.cpp [hunk 7]
```diff
@@ -164,33 +187,42 @@ InterimResult::buildIndex(const std::string &vidColumn) const {
...
+                    LOG(ERROR) << err;
+                    return Status::Error(err);
             }
         }
         index->rows_.emplace_back(std::move(row));
         ++rowIter;
     }
 
-    return index;
+    return std::move(index);
```
Comment (wadeliuyi) [inline src/graph/InterimResult.cpp]: return no need move, else it will effect RVO

File: src/graph/InterimResult.cpp [hunk 8]
```diff
@@ -369,13 +402,21 @@ std::unique_ptr<InterimResult> InterimResult::getInterim(
...
-    return std::make_unique<InterimResult>(std::move(rsWriter));
+    std::vector<std::string> colNames;
+    auto iter = resultSchema->begin();
+    while (iter) {
+        colNames.emplace_back(iter->getName());
+        ++iter;
+    }
+    auto result = std::make_unique<InterimResult>(std::move(colNames));
+    result->setInterim(std::move(rsWriter));
+    return std::move(result);
```
Comment (wadeliuyi) [inline src/graph/InterimResult.cpp]: ditto

File: src/graph/InterimResult.h [hunk 9]
```diff
@@ -41,22 +41,37 @@ class InterimResult final {
     static Status castToBool(cpp2::ColumnValue *col);
     static Status castToStr(cpp2::ColumnValue *col);
 
+    void setColNames(std::vector<std::string> colNames) {
```
Comment (wadeliuyi) [inline src/graph/InterimResult.h]: std::vector<std::string> &&colNames is better

File: src/graph/InterimResult.h [hunk 10]
```diff
@@ -41,22 +41,37 @@ class InterimResult final {
...
+            return nullptr;
+        }
         return rsReader_->schema();
     }
 
-    std::string& data() const {
-        return rsWriter_->data();
+    std::vector<std::string> getColNames() {
+        // Once getColNames called, colNames_ would be invalid
+        return std::move(colNames_);
```
Comment (wadeliuyi) [inline src/graph/InterimResult.h]: no need move

File: src/graph/OrderByExecutor.cpp [hunk 11]
```diff
@@ -136,8 +141,9 @@ void OrderByExecutor::execute() {
 }
 
 std::unique_ptr<InterimResult> OrderByExecutor::setupInterimResult() {
+    auto result = std::make_unique<InterimResult>(std::move(colNames_));
     if (rows_.empty()) {
-        return nullptr;
+        return std::move(result);
```
Comment (wadeliuyi) [inline src/graph/OrderByExecutor.cpp]: no need move

File: src/graph/OrderByExecutor.cpp [hunk 12]
```diff
@@ -167,23 +173,16 @@ std::unique_ptr<InterimResult> OrderByExecutor::setupInterimResult() {
         rsWriter->addRow(writer);
     }
 
-    return std::make_unique<InterimResult>(std::move(rsWriter));
+    result->setInterim(std::move(rsWriter));
+    return std::move(result);
```
Comment (wadeliuyi) [inline src/graph/OrderByExecutor.cpp]: ditto

File: src/graph/SetExecutor.cpp [hunk 13]
```diff
@@ -108,15 +110,31 @@ void SetExecutor::execute() {
...
             onError_(Status::Error(msg));
             return;
         }
 
-        if (leftResult_ == nullptr && rightResult_ == nullptr) {
+        if (leftResult_ == nullptr || rightResult_ == nullptr) {
+            // Should not reach here.
+            LOG(FATAL) << "Get null input: " << leftResult_.get()
+                        << " " << rightResult_.get();
+            onError_(Status::Error("Internal error."));
```
Comment (wadeliuyi) [inline src/graph/SetExecutor.cpp]: LOG(FATAL) will crash, onError_ will not execute, and indentation

File: src/graph/InterimResult.h [hunk 14]
```diff
@@ -41,22 +41,37 @@ class InterimResult final {
...
+
+    bool hasData() const {
+        return (rsWriter_ != nullptr) || (rsReader_ != nullptr);
+    }
+
     std::shared_ptr<const meta::SchemaProviderIf> schema() const {
+        if (!hasData()) {
+            return nullptr;
+        }
         return rsReader_->schema();
```
Comment (wadeliuyi) [inline src/graph/InterimResult.h]: when rsWriter_ != nullptr, but rsReader is nullptr, the hasData return true, but rsReader is nullptr, this call will crash

File: src/graph/InterimResult.cpp (hunk 6 above)
Comment (CPWstatic) [reply src/graph/InterimResult.cpp]: Good point.

File: src/graph/InterimResult.h (hunk 10 above)
Comment (wadeliuyi) [reply src/graph/InterimResult.h]: I have some test code, in this case it will call move constructor.

File: src/graph/SetExecutor.cpp (hunk 13 above)
Comment (wadeliuyi) [reply src/graph/SetExecutor.cpp]: the indentation after LOG(ERROR)

**Pull request merged**